protocols/noise/src/keys.rs

```diff
@@ -18,156 +18,185 @@
     }
 }
 
-impl AsRef<[u8]> for SecretKey {
-    fn as_ref(&self) -> &[u8] {
-        self.scalar.as_bytes()
+impl PartialEq for StaticPublicKey {
+    fn eq(&self, other: &StaticPublicKey) -> bool {
+        self.as_bytes() == other.as_bytes()
     }
 }
 
-/// ECC secret and public key.
-#[derive(Clone)]
-pub struct Keypair<T> {
-    secret: SecretKey,
-    public: PublicKey<T>
+impl Eq for StaticPublicKey {}
+
+impl From<[u8; 32]> for StaticPublicKey {
+    fn from(k: [u8; 32]) -> StaticPublicKey {
+        StaticPublicKey(x25519::PublicKey::from(k))
+    }
 }
 
-impl<T> Keypair<T> {
-    /// Create a new keypair.
-    pub fn new(s: SecretKey, p: PublicKey<T>) -> Self {
-        Keypair { secret: s, public: p }
+impl std::ops::Deref for StaticPublicKey {
+    type Target = x25519::PublicKey;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
     }
+}
 
-    /// Access this keypair's secret key.
-    pub fn secret(&self) -> &SecretKey {
-        &self.secret
+#[cfg(test)]
+mod tests {
+    use libp2p_core::identity::ed25519;
+    use quickcheck::*;
+    use sodiumoxide::crypto::sign;
+    use std::os::raw::c_int;
+    use super::*;
+
+    // ed25519 to x25519 keypair conversion must yield the same results as
+    // obtained through libsodium.
+    #[test]
+    fn prop_ed25519_to_x25519_matches_libsodium() {
+        fn prop() -> bool {
+            let ed25519 = ed25519::Keypair::generate();
+            let x25519 = StaticKeypair::from_ed25519(&ed25519);
+
+            let sodium_sec = ed25519_sk_to_curve25519(&sign::SecretKey(ed25519.encode()));
+            let sodium_pub = ed25519_pk_to_curve25519(&sign::PublicKey(ed25519.public().encode().clone()));
+
+            sodium_sec.as_ref() == Some(&x25519.secret.to_bytes()) &&
+            sodium_pub.as_ref() == Some(x25519.public.as_bytes())
+        }
+
+        quickcheck(prop as fn() -> _);
     }
 
-    /// Access this keypair's public key.
-    pub fn public(&self) -> &PublicKey<T> {
-        &self.public
+    // The x25519 public key obtained through ed25519 keypair conversion
+    // (and thus derived from the converted secret key) must match the x25519
+    // public key derived directly from the ed25519 public key.
+    #[test]
+    fn prop_public_ed25519_to_x25519_matches() {
+        fn prop() -> bool {
+            let ed25519 = ed25519::Keypair::generate();
+            let x25519 = StaticKeypair::from_ed25519(&ed25519);
+            let x25519_public = StaticPublicKey::from_ed25519(&ed25519.public());
+            Some(x25519.public()) == x25519_public.as_ref()
+        }
+
+        quickcheck(prop as fn() -> _);
     }
-}
 
-impl<T> Into<(SecretKey, PublicKey<T>)> for Keypair<T> {
-    fn into(self) -> (SecretKey, PublicKey<T>) {
-        (self.secret, self.public)
+    // Bindings to libsodium's ed25519 to curve25519 key conversions, to check that
+    // they agree with the conversions performed in this module.
+
+    extern "C" {
```

Comment:
I know it's a test, but by principle I'm not really fan of introducing any sort of unsafe call at all.

Reply:
What do you suggest? To remove the test?